### DISCUSSION TITLE
Lesson 6: Error while executing cast send foundry command

### DISCUSSION THREAD

**Discussion by albin-george-kurian** [Cyfrin/foundry-full-course-cu#878] [Q&A]
Getting Error: Failed to create wallet from private key. Private key is invalid hex: Odd number of digits
when executing command: cast send 0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0 "send(uint256)" 2388 --rpc-url $RPC_ID --private-key $PRIVATE_KEY

checked for,
1.Invisible white space(s)
2.Invalid private key
3.also updated `source .env` file after modifying

![Screenshot 2023-09-16 102719](https://github.com/Cyfrin/foundry-full-course-f23/assets/106237003/7b672d32-d780-43a9-9aa9-cfd93e8fb811)

@PatrickAlphaC

**Comment by theirrationalone** [comment]
You have an invalid privateKey in your .env file, Make sure there's no any white space before & after `=` operator. 

```env
PRIVATE_KEY=0xyour_valid_privateKey
RPC_URL=http://valid_rpc_url
```

**Reply by albin-george-kurian** [reply]
i think there is something else


![Screenshot 2023-09-16 150943](https://github.com/Cyfrin/foundry-full-course-f23/assets/106237003/a6977af0-702b-4c55-a95f-b86c29b63252)

**Reply by theirrationalone** [reply]
Invalid privateKey you have pasted maybe you accidently appended extra characters to your privateKey. 

Please replace your privateKey with valid one again.

**Reply by albin-george-kurian** [reply]
@theirrationalone ya i check it private key is correct

**Reply by theirrationalone** [reply]
Run anvil on a separate bash terminal, copy one of privateKeys offered by anvil again. Paste that into your .env file, save it. After that leave anvil running and open a new separated bash terminal and then deploy your contracts with --broadcast, rpc, privateKey subcommands then after, try again with your cast send.

**Reply by albin-george-kurian** [reply]
@theirrationalone it worked when hardcoded the value of private key into the command But didn't worked by using $PRIVATE_KEY 

![Screenshot 2023-09-16 155423](https://github.com/Cyfrin/foundry-full-course-f23/assets/106237003/63458d5a-863e-46d7-afa1-2ba41c8011e0)


But still I don't know why I did got that error while using $PRIVATE_KEY and still does getting that error while executing command with $PRIVATE_KEY

**Reply by theirrationalone** [reply]
The Problem might be extra trailing invisible white space(s) at the end of private key. Go back again to .env file and check it out if it's true.
You can confirm that by pressing `right arrow key` after placing cursor at the end of private key, if cursor shifts to right on pressing `right arrow key` then delete all trailing white spaces.

**Reply by albin-george-kurian** [reply]
spacings are correct


![Screenshot 2023-09-16 162729](https://github.com/Cyfrin/foundry-full-course-f23/assets/106237003/9bc93ceb-b5eb-4efb-9b31-ab678cbc78e8)

i tried replacing with other keys but still getting same error.

Hardcoding is working property: cast send 0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0 "send(uint256)" 2388  --rpc-url $RPC_ID --private-key 0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d

**Reply by theirrationalone** [reply]
Then last place of issue could be, You might be giving extra white space after putting `$PRIVATE_KEY`

Confirm it also. 

If above fails too, then there could have 3 possible issues

1. BUG
2. Invalid privateKey
3. Not updating `source .env` after modifying .env file

**Reply by albin-george-kurian** [reply]
updating?

**Reply by theirrationalone** [reply]
Execute bash cmd after updating   .env file

```bash
source  .env
```

**Reply by albin-george-kurian** [reply]
@theirrationalone not working..may be an bug or something..

**Reply by theirrationalone** [reply]
Try this

```bash
echo $PRIVATE_KEY
```

to see, How privateKey is being interpolated??...

**Reply by albin-george-kurian** [reply]
![Screenshot 2023-09-16 210123](https://github.com/Cyfrin/foundry-full-course-f23/assets/106237003/b23a5a84-038b-45ec-bfbf-f57f1c3be424)

**Reply by theirrationalone** [reply]
Definitely a BUG you've encountered Or it's something else we couldn't figured that out 😅

;`)

**Comment by albin-george-kurian** [comment]
@theirrationalone,
I have an another query when I tried to execute foundry related command in normal terminal present in vs code its not working,
But executing in wsl terminal is working properly...

have any solutions

![Screenshot 2023-09-16 204316](https://github.com/Cyfrin/foundry-full-course-f23/assets/106237003/1e2da07d-4c57-4057-85c0-413b1c0d51bb)

**Reply by theirrationalone** [reply]
You're not on a wsl instance, please click on botton left corner and start a wsl instance.

**Reply by theirrationalone** [reply]
There's no support for windows yet, that's why we are using wsl. 

Another thing you can do is, install bash on windows.

**Reply by theirrationalone** [reply]
You can download Git for Windows from the official website: [git](https://gitforwindows.org/) ... 

During installation, you can choose to enable Git Bash and optionally add it to your system PATH.

**Reply by albin-george-kurian** [reply]
👌👌

**Reply by theirrationalone** [reply]
@albin-george-kurian !, Pleasure to work with You ❤
